[fymbc] iP

fymbc opened the pull request:
## Friday

> “Why is Gamora?” – Drax ([source](https://www.youtube.com/watch?v=dQw4w9WgXcQ))

Friday ensures you get all things done before Friday!. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/fymbc/ip).
2. run the jar file.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

fymbc committed:
Rename Duke to Friday and add greeting

fymbc committed:
Add echo feature on Friday

fymbc committed:
Added Task and TaskList java classes

fymbc committed:
Implement basic exception handling for user input

fymbc committed:
Add completition status for tasks

fymbc committed:
Add task variants

fymbc committed:
Change TextUI test files

fymbc committed:
Added help menu and further error customisation

fymbc committed:
Enhance removal message and tracking of task numbers

fymbc committed:
Add persistent read write feature

fymbc committed:
Merge branch 'branch-Level-7'

fymbc committed:
Refactor code for more OOP

fymbc committed:
Merge branch 'branch-A-MoreOOP'

fymbc committed:
Refactor code into packages

fymbc committed:
Merge branch 'branch-A-Packages'

fymbc committed:
Fix minor display bugs in TaskList

fymbc committed:
Convert all dates from String to DateTime

fymbc committed:
Merge branch 'branch-Level-8'

fymbc committed:
Merge branch 'add-gradle-support'

fymbc committed:
Setup gradle to build and run project

fymbc committed:
Add JUnit tests for tasks

fymbc committed:
Merge branch 'branch-A-Junit'

wkxcass commented on src/test/java/tasks/TodoTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.jupiter.api.Test;
+import java.time.LocalDate;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps listing out the imported classes explicitly? I noticed similar issues in several other places too!

fymbc replied:
Hmm, I was following the notes which I believe imported it as a wildcard. I'll see how I can better list this. Thank you!

wkxcass commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,34 @@
+        this.done = false;
+    }
+
+    public String getName() { return this.name;
```
Perhaps moving the `return` statement into another line below?

fymbc replied:
Agreed! Thank you!

wkxcass commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+    public LocalDate getDate(String type, Scanner sc) {
+        try {
+            switch (type) {
+                case "d":
```
Perhaps removing the indentations for `case` clauses (as per the course's coding standards)? I noticed the same issue in several other places too!

fymbc replied:
Agreed! Thank you!

wkxcass commented on src/main/java/commands/removeCommand.java:
```diff
@@ -0,0 +1,26 @@
+package commands;
+import storage.Storage;
```
Perhaps adding a line between `package commands;` and `import storage.Storage;`? I found the same issue in several other places too!

fymbc replied:
Agreed! Thank you!

wkxcass commented on src/main/java/Friday.java:
```diff
@@ -0,0 +1,41 @@
+import commands.Command;
```
Perhaps putting the class into a package (as per the course's coding standards)? I would suggest creating a `friday` package, that would contain the `Friday` class as well as the other packages!

fymbc replied:
Agreed! Thank you!

wkxcass reviewed (commented):
Overall, I like how readable your code is! 👍
Just a couple of nits to fix.

fymbc committed:
Add JavaDocs to classes

fymbc committed:
Apply coding standard via checkstyle

fymbc committed:
Add keyword search function

fymbc committed:
Merge branch 'branch-A-JavaDoc'

fymbc committed:
Fix merge conflicts

fymbc committed:
Fix merge conflicts

fymbc committed:
Add JavaDocs for commands and fix style

cth06-Github commented on src/main/java/Friday.java:
```diff
@@ -0,0 +1,63 @@
+public class Friday {
+
+    private final Storage storage;
+    private final TaskList master;
```
May I check if there is any reason why "master" is used as the naming? Would "tasks" be a better naming as it conveys the meaning that it is about tasks?

fymbc replied:
No particular reason, just the first variable that came to find. Agreed it may be ambiguous. Have amended. Thank you!

cth06-Github commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class AddCommand implements Command {
+
+    private final String desc;
```
I believe that "desc" was written as a shortform of "description", but I think the shortform may be slightly ambigious, would keeping the full form be better? I think "description" is not a long word but I concede that this is a subjective matter!

fymbc replied:
Agreed! Have amended this for all commands. Thank you!

cth06-Github commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+     * @return false, indicating that the application should not terminate
+     */
+    @Override
+    public boolean execute(Storage storage, TaskList master, UI ui) {
```
May I check what this function does? My understanding is that the function manipulates the TaskList master and saves the list and print the line during execution, but after which it also returns whether it is executed. Would it be better to separate the functions into 2? We can have one `void execute()` function that does all the manipulation, and another boolean function to indicate whether the function is executed. This is to follow Single Responsibility Principle.

I notice that this function always return false, thus could I check what does this boolean status represent? I think The boolean function should ideally be named as `isExecuted()` or `hasStatus()` - with the prefix "is" or "has" followed by whatever the boolean status represent.

Same comment for `execute()` in other parts of the code.

fymbc replied:
The function was augmented to return a boolean function to distinguish between an exit command (which will return true) versus all other commands (which will return false), so that the program can catch and terminate when "bye" is inputted by the user. The aim is not to determine if command has execute successfully.

Noted on the Single Responsibility Principle. Let me try to think how I can improve this. Thank you!

cth06-Github commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+        String s = "";
+        for (int i = 0; i < master.getSize(); i++) {
+            Task task = master.getParent().get(i);
+            if (task.containsWord(this.desc)) {
```
Should this be `task.containWord(this.desc)`? (The verb "contains" should be in the base or root form.)

fymbc replied:
Thanks for catching this! Have amended.

cth06-Github commented on src/main/java/commands/MarkCommand.java:
```diff
@@ -0,0 +1,48 @@
+        try {
+            int index = Integer.parseInt(this.desc);
+
+            master.doneTask(this.action, index - 1);
```
May I clarify if `doneTask()` mean `markTaskAsDone()` or `showDoneTask()` (I.e. `showTaskDone()` -- show task that was done?

fymbc replied:
The former! Understand where the confusion may be. Have changed the naming. Thanks for pointing it out!

cth06-Github commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,55 @@
+        }
+        case "help" -> new HelpCommand();
+        default -> new InvalidCommand(parsed[0]);
+        };
```
I like how the arrow label way of writing switch - case block is used.

cth06-Github commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,161 @@
+        String[] values = line.split(",");
+        Task task;
+        switch (values[0]) {
+        case "td":
```
Would it be better if a comment is written to explain that case "td" means an instruction related to ToDo task? Same goes to "d" and "e".

fymbc replied:
Agreed! Have added further explanation in the JavaDocs. Thank you!

cth06-Github commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,161 @@
+        try (BufferedWriter bw = new BufferedWriter(new FileWriter(this.path))) {
+            for (Task task : list) {
+                String str = "";
+                if (task instanceof Todo) {
+                    str = "td," + task.getName();
+                    if (task.isDone()) {
+                        str += ",y";
+                    } else {
+                        str += ",n";
+                    }
+                }
+                if (task instanceof Deadline d) {
+                    str += "d," + d.getName();
+                    if (task.isDone()) {
+                        str += ",y,";
+                    } else {
+                        str += ",n,";
+                    }
+                    str += d.getDeadline();
+                }
+                if (task instanceof Event e) {
+                    str += "e," + e.getName();
+                    if (task.isDone()) {
+                        str += ",y,";
+                    } else {
+                        str += ",n,";
+                    }
+                    str += e.getStart() + "," + e.getEnd();
+                }
```
There seem to be repeated lines of code. If these functions are written as part of the Todo, Deadline, Event class using the same name, then we can make use of polymorphism and the output of the string to be saved into Storage will automatically change without the need for `instanceof`. It will be beneficial because if we need to create another new class that inherits from Task, we do not have to add in 
```Java
if (instanceof newClass n) {
    // code on how the string should be printed/saved into the text file.
} 
``` 
to give an example.

Definitely feel free to disagree because code design wise...while I think this suggestion will make the code better, maybe there are problems I did not notice such that "this functionality should not be written in this class but another class".

fymbc replied:
You have a point. The motivation is to parse it into a CSV-esque format that reduces potential errors in parsing.

Perhaps I can shift this "to-CSV-format" functionality into the tasks itself and make use of polymorphism.

Thank you!

cth06-Github commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+ * The TaskList class is responsible for handling all operations on the tasks stored in the list.
+ */
+public class TaskList {
+    private final ArrayList<Task> parent;
```
Would it be better if the variable is named as tasks? (or anything else that is plural?) I am guessing that "parent" was used to indicate that under the hood, or, at its most fundamental level, the TaskList class is like an `ArrayList<Task>` ? So it does feel like `ArrayList<Task>` is the parent? I guess TaskList and ArrayList are associated but they aren't parent-child relationship as if it was inheritance, so it might be a misnomer. Moreover I think it's more important to show what does this attribute do / mean, rather than showing the relationship or association, If my speculation on why the word "parent" is used is true.

fymbc replied:
Agreed! Amended! Thank you!

cth06-Github commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+     * @param action the action to perform ("mark" for done, any other value for undone)
+     * @param task the index of the task to be marked
+     */
+    public void doneTask(String action, int task) {
```
Would `markTaskAsDone()` be better, although the name is longer? In one of the above comments, I asked whether it should be `markTaskAsDone()` or `showDoneTask()` as I feel that it is not exactly clear whether "done" in this case is meant to be a verb but written in wrong tense or meant to describe the task (adjective).

fymbc replied:
I agree! Noted and amended. Thanks for catching it!

cth06-Github commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+            switch (type) {
+            case "d":
+                return LocalDate.parse(getInputFromUser(sc, "Deadline (in dd mm yyyy) > "), this.inputFormatter);
+            case "es":
```
What does "es" stand for?

cth06-Github commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+                return LocalDate.parse(getInputFromUser(sc, "Deadline (in dd mm yyyy) > "), this.inputFormatter);
+            case "es":
+                return LocalDate.parse(getInputFromUser(sc, "Start Date (in dd mm yyyy) > "), this.inputFormatter);
+            case "ee":
```
What does "ee" stand for?

fymbc replied:
Event end! Which denotes the date expected from the input should be the event end date.

cth06-Github commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+     * @param test true to count completed tasks, false to count incomplete tasks
+     * @return the number of tasks that match the specified completion status
+     */
+    public int countCompleted(boolean test) {
```
```suggestion
    public int countCompleted(boolean isTested) {
```

cth06-Github replied:
Would it be possible to clarify what boolean test variable represents? I'm not sure whether my interpretation is correct.

fymbc replied:
It's supposed to be a dual use function, where it returns the number of completed tasks if boolean is set to true, and number of uncompleted tasks if boolean is set to false. But I see where the confusion may be, have made amendments. Thank you!

cth06-Github commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return the start date of the event
+     */
+    public LocalDate getStart() {
```
```suggestion
    public LocalDate getStartDate() {
```

fymbc replied:
Thank you for catching this!

cth06-Github commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return the end date of the event
+     */
+    public LocalDate getEnd() {
```
```suggestion
    public LocalDate getEndDate() {
```

fymbc replied:
Thank you for catching this!

cth06-Github commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public abstract class Task {
+    private final String name;
+    private boolean done;
```
```suggestion
    private boolean isDone;
```

fymbc replied:
Thank you for catching this!

cth06-Github reviewed (commented):
Good work, code naming mostly adhered to! I hope the review comments are alright!

fymbc committed:
Syntax correction for TaskList

fymbc committed:
Add JavaFX features

fymbc committed:
Merge branch 'branch-Level-10'

fymbc committed:
Add some assertions to verify program integrity

fymbc committed:
Output from Friday: Resolve UI bugs

After refactoring Friday into a separate package,
output has changed from Friday to Friday.Friday.

Changing it back to only Friday can improve neatness and readability.

Let's modify the outputs from System.out to show Friday instead of
Friday.Friday.

fymbc committed:
Merge branch 'master' of https://github.com/fymbc/ip into branch-A-Assertions

fymbc committed:
Merge pull request #1 from fymbc/branch-A-Assertions

Add assertion functionality to Friday.java and TaskList.java

fymbc committed:
Remove unused variables and change access modifiers

Current method signatures for commands have unused variables,
and some classes' fields can also be made final.

Removing unused variables can make the code cleaner
and setting 'final' for fields can take advantage of immutability.

Let's update these classes to have final variables while removing
unused variables

fymbc committed:
Fix keyword search functionality

fymbc committed:
Merge branch 'master' of https://github.com/fymbc/ip into branch-A-CodeQuality

fymbc committed:
Improve code quality and readability

Current code has poorly named methods and fields.

This can impact readability and understanding due to
potential ambiguity.

Let's make it clearer by changing some of the methods and fields
to a clearer name in accordance to PR suggestions and coding style guide

fymbc committed:
Fix code quality issues

fymbc committed:
Merge pull request #3 from fymbc/branch-A-CodeQuality

Improve code quality for methods and fields naming and headers

fymbc committed:
Add yml for CI

fymbc committed:
Remove Gradle checkstyle

fymbc committed:
Add functionality for checking duplicates

fymbc committed:
Merge branch 'branch-BCD-Extension'

fymbc committed:
Refactor code for more OOP

fymbc committed:
Refactor code for more OOP

fymbc committed:
Merge branch 'branch-A-CodeQuality'

fymbc committed:
Fix minor checkstyle issues

fymbc committed:
Refactor code for JavaFX compatability

Current code relies on printing to console for
user interaction. This implies imcomptability with
JavaFX as it relies on String responses from user commands.

Let's change it to String formatted output

fymbc committed:
Refine compatibility for JavaFX

fymbc committed:
Improve JavaFX GUI to Level 5

fymbc committed:
Add more robust error handling and fix minor bugs

fymbc committed:
Add minor GUI enhancements

fymbc committed:
Update user guide
